codex-rs/core/src/shell.rs

```diff
@@ -131,6 +134,178 @@ fn strip_bash_lc(command: &Vec<String>) -> Option<String> {
+fn extract_shell_script(command: &[String]) -> Option<(usize, &str)> {
+    match command {
+        [shell, flag, script]
+            if looks_like_shell_wrapper(shell) && (flag == "-lc" || flag == "-c") =>
+        {
+            Some((2, script.as_str()))
+        }
+        _ => None,
+    }
+}
+
+fn looks_like_shell_wrapper(shell: &str) -> bool {
+    let name = Path::new(shell)
+        .file_name()
+        .and_then(|s| s.to_str())
+        .unwrap_or(shell);
+
+    matches!(name, "bash" | "sh" | "zsh")
+}
+
+/// Result of stripping a single leading `cd` command from an interactive shell
+/// script. The command `cd <dir> && rest` yields the directory (kept for future
+/// use) alongside the remainder of the script.
+struct ShellScriptCdSplit {
+    _workdir: String,
+    remainder: String,
+}
+
+fn split_cd_prefix_from_script(script: &str) -> Option<ShellScriptCdSplit> {
+    let tree = try_parse_bash(script)?;
+    let mut commands = try_parse_word_only_commands_sequence(&tree, script)?;
+    if commands.is_empty() {
+        return None;
+    }
+
+    let cd_command = commands.pop()?;
+    if cd_command.first().map(String::as_str) != Some("cd") {
+        return None;
+    }
+
+    if cd_command.len() != 2 {
+        return None;
+    }
+
+    if commands
+        .iter()
+        .any(|cmd| cmd.first().map(String::as_str) == Some("cd"))
+    {
+        return None;
+    }
+
+    let workdir = cd_command.get(1)?.clone();
+    if workdir.is_empty() {
+        return None;
+    }
+
+    let tokens = shlex::split(script)?;
+    let rest_tokens = tokens_after_first_command(&tokens)?;
+    let rest_script = join_shell_words_preserving_operators(rest_tokens)?;
+
+    Some(ShellScriptCdSplit {
+        _workdir: workdir,
+        remainder: rest_script,
+    })
+}
+
+/// Returns the tokens that follow the first command when it is chained with
+/// `&&`. If the script does not start with `cd <dir> && ...`, `None` is
+/// returned.
+fn tokens_after_first_command(tokens: &[String]) -> Option<&[String]> {
+    let mut idx = 0;
+    while idx < tokens.len() && tokens[idx] != "&&" {
+        idx += 1;
+    }
+
+    if idx == tokens.len() {
+        return None;
+    }
+
+    if tokens[idx] != "&&" {
+        return None;
+    }
+
+    let rest_start = idx + 1;
+    if rest_start >= tokens.len() {
+        return None;
+    }
+
+    Some(&tokens[rest_start..])
+}
+
+fn join_shell_words_preserving_operators(tokens: &[String]) -> Option<String> {
+    if tokens.is_empty() {
+        return None;
+    }
+
+    let mut parts = Vec::with_capacity(tokens.len());
+    for token in tokens {
+        if is_shell_operator(token) {
+            parts.push(token.clone());
+        } else {
+            let quoted = shlex::try_join(std::iter::once(token.as_str())).ok()?;
+            parts.push(quoted);
+        }
+    }
+
+    Some(parts.join(" "))
+}
+
+fn is_shell_operator(token: &str) -> bool {
+    matches!(token, "&&" | "||" | "|" | ";")
```

Comment:
I'm not sure what this list means, but I don't think `|` should be in this list.